### ISSUE TITLE
Make fails on version 4.2.1 but not 4.1

### ISSUE THREAD

**Issue by infomaniac50** [eerotal/LibreSignage#48]
I can't say for sure but I think make is bugging out on version 4.2.1. This is the version that I have on ArchLinux. To compare, I made a Docker in Docker build using Ubuntu:18.04 and everything worked fine.

## TLDR

On ArchLinux make 4.2.1 says:
``dep/logout/js/main.js.dep:3: *** target pattern contains no '%'.  Stop.``

Read the rest for more context.

### ArchLinux
**[signage-experimental/LibreSignage] $ git branch**
```
  master
* next
```

**[signage-experimental/LibreSignage] $ make --version**
```
GNU Make 4.2.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**[signage-experimental/LibreSignage] $ node --version**
``v6.17.0``

**[signage-experimental/LibreSignage] $ npm --version**
``6.9.0``

**[signage-experimental/LibreSignage] $ npx --version**
``6.9.0``

**[signage-experimental/LibreSignage] $ gcc --version**
```
gcc (GCC) 8.2.1 20181127
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**[signage-experimental/LibreSignage] $ g++ --version**
```
g++ (GCC) 8.2.1 20181127
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**[signage-experimental/LibreSignage] $ docker version**
```
Client:
 Version:           18.09.3-ce
 API version:       1.39
 Go version:        go1.12
 Git commit:        774a1f4eee
 Built:             Thu Feb 28 20:38:40 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.3-ce
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.12
  Git commit:       774a1f4eee
  Built:            Thu Feb 28 20:37:49 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

### DinD Ubuntu 18.04

**root@b15331019df3:/LibreSignage# git branch**
```
* next
```

**root@b15331019df3:/LibreSignage# make --version**
```
GNU Make 4.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```
**root@b15331019df3:/LibreSignage# nodejs --version**
```
v8.10.0
```

**root@b15331019df3:/LibreSignage# npm --version**
```
6.9.0
```

**root@b15331019df3:/LibreSignage# npx --version**
```
10.2.0
```

**root@b15331019df3:/LibreSignage# gcc --version**
```
gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**root@b15331019df3:/LibreSignage# g++ --version**
```
g++ (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**root@b15331019df3:/LibreSignage# cat /etc/lsb-release**
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
```

**root@b15331019df3:/LibreSignage# docker version**
```
Client:
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        6247962
 Built:             Tue Feb 26 23:52:23 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.3-ce
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.12
  Git commit:       774a1f4eee
  Built:            Thu Feb 28 20:37:49 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```
### Dockerfile.builder

This file goes in the same directory as Dockerfile.

```Dockerfile
FROM ubuntu:18.04
RUN DEBIAN_FRONTEND=noninteractive apt-get update
RUN DEBIAN_FRONTEND=noninteractive apt-get upgrade -y
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y build-essential git apache2 php php-gd pandoc make rsync curl docker-compose imagemagick
RUN DEBIAN_FRONTEND=noninteractive curl -sL https://deb.nodesource.com/setup_6.x | bash -
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y nodejs npm
RUN DEBIAN_FRONTEND=noninteractive npm -g install npm npx
RUN ln -fs /usr/share/zoneinfo/UTC /etc/localtime
RUN dpkg-reconfigure --frontend noninteractive tzdata
```

To build and start the builder image, use these shell commands on the host.

```bash
docker build --file Dockerfile.builder --tag libresignage-builder .
docker create --privileged --mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock --interactive --tty --name libresignage-builder libresignage-builder
docker start --interactive --attach libresignage-builder
```

## Notes on Docker in Docker

There was some nonsense about invalid pointers being freed when I used my Docker in Docker config. I'm probably doing something wrong here. I need to read up on Docker in Docker to make sure I'm doing everything correctly.

```
free(): invalid pointer
SIGABRT: abort
PC=<redacted hex address> m=0 sigcode=<redacted 64 bit number>
signal arrived during cgo execution
```

## Command History & Output on ArchLinux

**[signage-experimental/LibreSignage] next $ npm install**
```
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.7 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.7: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 388 packages from 274 contributors and audited 6253 packages in 18.19s
found 0 vulnerabilities
```

**[signage-experimental/LibreSignage] next $ make configure TARGET=apache2-debian-docker PASS="--features imgthumbs,vidthumbs"**
```
[INFO] Configuring for 'apache2-debian-docker'.
[INFO] Pass to build config generator: '--features imgthumbs,vidthumbs'
Write config to 'build/docker_2019-03-18_20:32:03.conf'.
[Info] Load config from 'signage-experimental/LibreSignage/build/docker_2019-03-18_20:32:03.conf'
[Info] Load config from 'signage-experimental/LibreSignage/build/docker_2019-03-18_20:32:03.conf'
```
**[signage-experimental/LibreSignage] next $ make**
```
[Info] Include dependency makefiles.
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/app/css/display.scss >> dep/app/css/display.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/control/usermgr/css/usermgr.scss >> dep/control/usermgr/css/usermgr.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/control/css/control.scss >> dep/control/css/control.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/control/editor/css/main.scss >> dep/control/editor/css/main.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/control/user/css/user.scss >> dep/control/user/css/user.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/errors/css/error.scss >> dep/errors/css/error.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/doc/css/doc.scss >> dep/doc/css/doc.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/login/css/login.scss >> dep/login/css/login.scss.dep
/bin/sh: line 4: warning: command substitution: ignored null byte in input
deps-scss: src/logout/css/logout.scss >> dep/logout/css/logout.scss.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/app/js/main.js >> dep/app/js/main.js.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/control/usermgr/js/main.js >> dep/control/usermgr/js/main.js.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/control/js/main.js >> dep/control/js/main.js.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/control/editor/js/main.js >> dep/control/editor/js/main.js.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/control/user/js/main.js >> dep/control/user/js/main.js.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/doc/js/main.js >> dep/doc/js/main.js.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/login/js/main.js >> dep/login/js/main.js.dep
/bin/sh: line 2: warning: command substitution: ignored null byte in input
deps-js: src/logout/js/main.js >> dep/logout/js/main.js.dep
[Info] Include dependency makefiles.
dep/logout/js/main.js.dep:3: *** target pattern contains no '%'.  Stop.
```

## Command History & Output on DinD Ubuntu 18.04

**root@b15331019df3:/LibreSignage# npm install**
```
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.7 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.7: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 388 packages from 274 contributors and audited 6253 packages in 16.022s
found 0 vulnerabilities
```

**root@b15331019df3:/LibreSignage# make configure TARGET=apache2-debian-docker PASS="--features imgthumbs,vidthumbs"**
```
[INFO] Configuring for 'apache2-debian-docker'.
[INFO] Pass to build config generator: '--features imgthumbs,vidthumbs'
Write config to 'build/docker_2019-03-19_01:28:55.conf'.
[Info] Load config from '/LibreSignage/build/docker_2019-03-19_01:28:55.conf'
[Info] Load config from '/LibreSignage/build/docker_2019-03-19_01:28:55.conf'
```
**root@b15331019df3:/LibreSignage# make**
*Long build output skipped*
*Build was successful*

**root@b15331019df3:/LibreSignage# make install**
```
[Info] Include dependency makefiles.
[Info] Load config from '/LibreSignage/build/docker_2019-03-19_01:28:55.conf'
[Info] Load config from '/LibreSignage/build/docker_2019-03-19_01:28:55.conf'
[INFO] Args for 'docker build': --build-arg version=v1.0.0-beta-1-38-gbc3b1ca --build-arg imgthumbs=y --build-arg vidthumbs=y
```
*Long build output skipped*
```
Successfully built 641e5ccb3d14
Successfully tagged libresignage:v1.0.0-beta-1-38-gbc3b1ca
```

**Comment by infomaniac50** [comment]
Technically it's not Docker in Docker. I'm using bind mounts to access the host's Docker socket file. /var/run/docker.sock Doing it this way keeps me from having to bootstrap a full VM in VirtualBox.

It seems I did do it correctly so the SIGABRT stuff is probably something complaining but not really having a detrimental effect.

https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/

**Comment by eerotal** [comment]
This seems to be the same issue as #34, which I haven't been able to debug because of insufficient information. Your issue has every little detail included so that's great. 

I have encountered cases where make has been picky about makefile syntax when switching versions and this looks like a similar issue. I'll take a closer look at this when I get on my PC the next time.

**Comment by eerotal** [comment]
In fact, the problem doesn't seem to be with the make versions. I run make 4.2.1 on my system (Debian Buster) and I can build LibreSignage just fine. Could you attach the contents of *dep/logout/js/main.js.dep* here after trying to build on Arch Linux? You could also try running `make --trace` to see where exactly the problem is. The output is **very** verbose, so you can just attach a part of it here. I think this problem has something to do with some odd shell script syntax problem related to the warning message `/bin/sh: line 4: warning: command substitution: ignored null byte in input`.

**Comment by infomaniac50** [comment]
It seems the tab escape character isn't being interpreted.

[make-trace.txt](https://github.com/eerotal/LibreSignage/files/2993977/make-trace.txt)
[main.js.dep.txt](https://github.com/eerotal/LibreSignage/files/2993978/main.js.dep.txt)

There was some weirdness with my shell startup files so as a precaution I tried to build it in an ArchLinux container. However, the result was the same.

[main.js.dep-arch-docker.txt](https://github.com/eerotal/LibreSignage/files/2994101/main.js.dep-arch-docker.txt)
[make-trace-arch-docker.txt](https://github.com/eerotal/LibreSignage/files/2994102/make-trace-arch-docker.txt)

Here is the build files from my Ubuntu 18.04 container.
[main.js.dep-ubuntu.txt](https://github.com/eerotal/LibreSignage/files/2994144/main.js.dep-debian.txt)
[make-trace-ubuntu.txt](https://github.com/eerotal/LibreSignage/files/2994145/make-trace-debian.txt)

**Comment by eerotal** [comment]
I think I tracked down the issue. The calls to `echo` that generate the *.dep files are missing the `-e` flag which enables interpretation of backslash escapes. Now, interestingly, the code works fine on Debian and Ubuntu, so there must be some differences between `echo` on different shells. I'll fix this when I get on my PC.

**Comment by eerotal** [comment]
The `** target pattern contains no '%'. Stop.` error and the `/bin/sh: line 4: warning: command substitution: ignored null byte in input` were indeed two separate problems. They both were caused by the same thing though: `/bin/sh` seems to actually be bash on your system. By default make should be using `sh` but if `/bin/sh` is something else, it doesn't even know what shell it's running. The makefile should now be more robust against cases like this.

**Comment by infomaniac50** [comment]
It compiled successfully on my host system. It compiled successfully in Docker too.

I saw that make used sh so I decided to see what package owned sh. Sure enough /bin/sh is symlinked to bash. I didn't know if that had any effect on the scripts though.

The Arch Wiki says, "In Arch ``/bin/sh`` (which used to be the Bourne shell executable) is symlinked to ``/bin/bash``. If Bash is invoked with the name ``sh``, it tries to mimic the startup behavior of historical versions of ``sh``, including POSIX compability."

```
$ pacman -Qo /bin/sh
/usr/bin/sh is owned by bash 5.0.002-1
```
